### PR TITLE
Permutation matrix

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

#1686

#### Other comments

I've reworked #1686 to use Permutation, but this is still work in progress and I would want to check the coverage here.

There was a need to allow the `Permutation` to become symbolic functions because some evaluations build dummy summations in the intermediate, but I have done via adding a new object `AppliedPermutation` rather than using the old function framework.

I've also found the `Permutation` applied for negative integers, so I have disabled it.
And I have made `Permutation.apply` to use it like a function that takes over a single variable.

I've added `Permutation.resize` that can perform on resizing permutation by dropping some cycles.

I've also added `MatrixPermute` class which can be symbolic analogous to `Matrix.permute`.

#### Release Notes

- combinatorics
  - Fixed `Permutation` applying for negative integers.
  - Added `Permutation.apply` to use permutation as a symbolic function.

- matrices
  - Added `PermutationMatrix` and `MatrixPermute` for matrix permutations.

### CODE REVIEW

File: sympy/combinatorics/permutations.py [hunk 1]
```diff
@@ -2832,6 +2887,72 @@ def unrank_lex(cls, size, rank):
             psize = new_psize
         return cls._af_new(perm_array)
 
+    def resize(self, n):
+        """Resize the permutation to the new size $n$.
```
Comment (oscarbenjamin) [inline sympy/combinatorics/permutations.py]: In the new style guide parameters are referred to in asterisks e.g. `*n*`

File: sympy/combinatorics/permutations.py [hunk 2]
```diff
@@ -1623,6 +1629,55 @@ def atoms(self):
...
+            have integer values, an ``AppliedPermutation`` object
+            will be returned which can represent an unevaluated
+            function.
+
+        Notes
+        =====
+
+        Any permutation can be defined as a bijective function
+        `\sigma : \{ 0, 1, ..., n-1 \} \rightarrow \{ 0, 1, ..., n-1 \}`
+        where `n` denotes the size of the permutation.
```
Comment (oscarbenjamin) [inline sympy/combinatorics/permutations.py]: Should these be enclosed in `$` rather than backticks?

**Comment by oscarbenjamin** [comment]
Looks good to me

**Comment by oscarbenjamin** [comment]
There are a few lines not covered by the tests (according to codecov)

**Comment by oscarbenjamin** [comment]
+1 to merge (Travis is busy right now so I'll wait a bit...)

**Comment by sylee957** [comment]
It still needs some work for adding test cases for MatrixPermute

File: sympy/matrices/expressions/permutation.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+    _eval_transpose = _eval_adjoint = _eval_inverse
+
+    def _eval_determinant(self):
+        sign = self.args[0].signature()
+        if sign == 1:
+            return S.One
+        elif sign == -1:
+            return S.NegativeOne
+        raise NotImplementedError
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/permutation.py]: This is the only line not covered by the tests. Can the sign be anything other than 1 or -1?

Comment (sylee957) [reply sympy/matrices/expressions/permutation.py]: It won't likely be other than `1` or `-1`

File: sympy/matrices/expressions/tests/test_permutation.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        mat = PermutationMatrix(perm)
+        inv = Inverse(mat).as_explicit()
+        assert inv == mat.as_explicit().inv()
+
+
+def test_PermutationMatrix_rewrite_BlockDiagMatrix():
+    for perm in SymmetricGroup(6).generate():
+        mat = PermutationMatrix(perm)
+        rewrite = mat.rewrite(BlockDiagMatrix)
+        assert mat.as_explicit() == rewrite.as_explicit()
```
Comment (sylee957) [inline sympy/matrices/expressions/tests/test_permutation.py]: This test takes a lot of time because I have not provided limited set of examples.

**Pull request merged**